File: app/src/main/java/nl/eduvpn/app/entity/DiscoveredAPI.java
/*
 *  This file is part of eduVPN.
 *
 *     eduVPN is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     eduVPN is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with eduVPN.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.eduvpn.app.entity;

import android.net.Uri;

import androidx.annotation.NonNull;
import nl.eduvpn.app.Constants;

/**
 * A discovered API entity, containing all the URLs.
 * Created by Daniel Zolnai on 2016-10-18.
 */
public class DiscoveredAPI {

    private String _apiBaseUri;
    private String _authorizationEndpoint;
    private String _tokenEndpoint;

    public DiscoveredAPI(@NonNull String apiBaseUri,
                         @NonNull String authorizationEndpoint,
                         @NonNull String tokenEndpoint) {
        _apiBaseUri = apiBaseUri;
        _authorizationEndpoint = authorizationEndpoint;
        _tokenEndpoint = tokenEndpoint;
    }

    @NonNull
    public String getAuthorizationEndpoint() {
        return _authorizationEndpoint;
    }

    public String getApiBaseUri() {
        return _apiBaseUri;
    }

    public String getTokenEndpoint() {
        return _tokenEndpoint;
    }

    @NonNull
    public String getSystemMessagesEndpoint() {
        return Uri.parse(_apiBaseUri).buildUpon().appendPath(Constants.API_SYSTEM_MESSAGES_PATH).build().toString();
    }

    @NonNull
    public String getUserMessagesEndpoint() {
        return Uri.parse(_apiBaseUri).buildUpon().appendPath(Constants.API_USER_MESSAGES_PATH).build().toString();
    }

    @NonNull
    public String getProfileListEndpoint() {
        return Uri.parse(_apiBaseUri).buildUpon().appendPath(Constants.API_PROFILE_LIST_PATH).build().toString();
    }

    @NonNull
    public String getCreateKeyPairEndpoint() {
        return Uri.parse(_apiBaseUri).buildUpon().appendPath(Constants.API_CREATE_KEYPAIR).build().toString();
    }

    @NonNull
    public String getProfileConfigEndpoint() {
        return Uri.parse(_apiBaseUri).buildUpon().appendPath(Constants.API_PROFILE_CONFIG).build().toString();
    }

    @NonNull
    public String getCheckCertificateEndpoint(String certCommonName) {
        return Uri.parse(_apiBaseUri).buildUpon()
                .appendPath(Constants.API_CHECK_CERTIFICATE)
                .appendQueryParameter("common_name", certCommonName)
                .build()
                .toString();
    }
}

File: app/src/main/java/nl/eduvpn/app/entity/DiscoveredAPI.kt
/*
 *  This file is part of eduVPN.
 *
 *     eduVPN is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     eduVPN is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with eduVPN.  If not, see <http://www.gnu.org/licenses/>.
 */
package nl.eduvpn.app.entity

import android.net.Uri
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import nl.eduvpn.app.Constants

/**
 * A discovered API entity, containing all the URLs.
 * Created by Daniel Zolnai on 2016-10-18.
 */

sealed class DiscoveredAPI {
    abstract val authorizationEndpoint: String
    abstract val tokenEndpoint: String

    fun toDiscoveredAPIs(): DiscoveredAPIs {
        return when (this) {
            is DiscoveredAPIV2 -> DiscoveredAPIs(this, null)
            is DiscoveredAPIV3 -> DiscoveredAPIs(null, this)
        }
    }

    internal fun getURL(baseURI: String, path: String): String =
        Uri.parse(baseURI)
            .buildUpon()
            .appendPath(path)
            .build()
            .toString()
}

@Serializable
class DiscoveredAPIV3(

    @SerialName("api_endpoint")
    val apiEndpoint: String,

    @SerialName("authorization_endpoint")
    override val authorizationEndpoint: String,

    @SerialName("token_endpoint")
    override val tokenEndpoint: String
) : DiscoveredAPI() {

    val infoEndpoint: String by lazy {
        getURL(apiEndpoint, Constants.API_V3_INFO_PATH)
    }

    val connectEndpoint: String by lazy {
        getURL(apiEndpoint, Constants.API_V3_CONNECT_PATH)
    }

    val disconnectEndpoint: String by lazy {
        getURL(apiEndpoint, Constants.API_V3_DISCONNECT_PATH)
    }
}

@Serializable
class DiscoveredAPIV2(
    @SerialName("api_base_uri")
    val apiBaseUri: String,

    @SerialName("authorization_endpoint")
    override val authorizationEndpoint: String,

    @SerialName("token_endpoint")
    override val tokenEndpoint: String
) : DiscoveredAPI() {

    val profileListEndpoint: String by lazy {
        getURL(apiBaseUri, Constants.API_PROFILE_LIST_PATH)
    }

    val createKeyPairEndpoint: String by lazy {
        getURL(apiBaseUri, Constants.API_CREATE_KEYPAIR)
    }

    val profileConfigEndpoint: String by lazy {
        getURL(apiBaseUri, Constants.API_PROFILE_CONFIG)
    }

    fun getCheckCertificateEndpoint(certCommonName: String?): String {
        return Uri.parse(apiBaseUri).buildUpon()
            .appendPath(Constants.API_CHECK_CERTIFICATE)
            .appendQueryParameter("common_name", certCommonName)
            .build()
            .toString()
    }
}
